src/main/java/rx/internal/operators/OperatorConcat.java

```diff
@@ -212,6 +227,11 @@ public void onError(Throwable e) {
         @Override
         public void onCompleted() {
             if (once.compareAndSet(0, 1)) {
+                // signal the production count at once instead of one by one
+                long c = produced;
+                if (c != 0L) {
```

Comment:
Because of the compareAndSet, the read of parent field would happen all the time, including empty sources. This also affects produced which may or may not be register-cached.